packages/dds/tree/src/core/tree/anchorSet.ts

```diff
@@ -33,7 +35,106 @@ export interface AnchorLocator {
 	 * for now caller must rely on data in anchor + returned node location
 	 * (not ideal for anchors for places or ranges instead of nodes).
 	 */
-	locate(anchor: Anchor): UpPath | undefined;
+	locate(anchor: Anchor): AnchorNode | undefined;
+}
+
+/**
+ * @alpha
+ */
+export type AnchorKeyBrand = Brand<number, "AnchorKey">;
+
+/**
+ * @alpha
+ */
+export type AnchorKey<TContent> = Opaque<AnchorKeyBrand> & Covariant<TContent>;
+let slotCounter = 0;
+
+/**
+ * Data stashed on an anchor.
+ * @alpha
+ */
+export interface MapSubset<K, V> {
+	get(key: K): V | undefined;
+	has(key: K): boolean;
+	set(key: K, value: V): this;
+	delete(key: K): boolean;
+}
+
+/**
+ * Events for {@link AnchorNode}.
+ *
+ * TODO:
+ * - Design how events should be ordered.
+ * - Determine if events should be deferred until update of forest is done, and/or AnchorSet and forest should be updated together with events as it goes.
+ * - Add more events.
+ * - Work out how slots should interact with event related lifetime extension.
+ *
+ * @alpha
+ */
+export interface AnchorEvents {
+	/**
+	 * When the anchor node will never get reused by its AnchorSet.
+	 * This means that the content it corresponds to has been deleted, and that if its undeleted it will be treated as a recreation.
+	 *
+	 * This may happen after a delay after the content was removed from the tree, immediately, or never: it depends on how the AnchorSet is being used.
+	 */
+	afterDelete(anchor: AnchorNode): void;
+}
+
+/**
+ * Node in an tree of anchors.
+ * @alpha
+ */
+export interface AnchorNode extends UpPath<AnchorNode>, ISubscribable<AnchorEvents> {
+	/**
+	 * Allows access to data stored on the Anchor in "slots".
+	 * Use {@link anchorSlot} to create slots.
+	 *
+	 * @remarks
+	 * This is not fully type safe.
+	 */
+	slotMap<T extends AnchorKey<unknown>>(
+		slot: T,
+	): T extends AnchorKey<infer TContent>
+		? MapSubset<T, TContent>
+		: Map<AnchorKey<unknown>, unknown>;
+
+	/**
+	 * Gets a child of this node.
+	 *
+	 * @remarks
+	 * This does not return an AnchorNode since there might not be one, and lazily creating one here would have messy lifetime management.
+	 * If an AnchorNode is requires, use the AnchorSet to track then locate the returned path.
```

Comment:
if the client exists, then you are holding onto it without a refcount, and it might vanish on you. If the client doesn't exist, then you have to create it and add a ref, which would cause a leak. We could unconditionally ad a ref when you access a child, but I suspect most users of such an API would get it wrong, and doing refs like that manually would defeat any tooling we add to help catch such errors. Instead you get an Anchor which corosponds to a ref counted ref using track (which just calls child and adds a ref in this case) which means that you can tell if you have leaked anchors by lookinf for unfreed entries from track.

In a world with recounting smart pointers things would be different, but implementing those in typescript doesn't really work well.